Contextual mode

Arantor opened the pull request:
This needs so much testing. **So so much testing.** And even after it's thoroughly tested, it needs more testing. I suspect there will be edge cases against replying to your own topics, and attachments are a level of hellspawn. As is post moderation. As is editing posts. As is editing posts-that-legitimately-no-longer-have-an-author-account. But it _seems_ to work.

Fixes #448
Should fix #242 
Should fix #378
Should fix #370 
Should fix #188

Arantor committed:
Some settings for contextual mode.

Arantor committed:
Make sure sane defaults exist; most settings are relatively mild if absent or incorrect; this has the potential to break a lot of things.

Arantor committed:
Implement immersive mode/non immersive mode for overall groups.

Arantor committed:
First pass at a characters-allowed-to function, taking into account that characters don't get basically any of the membergroup level permissions.

Arantor committed:
Merge branch 'master' of https://github.com/SockDrawer/StoryBB into contextual-mode

Arantor committed:
Merge branch 'master' of https://github.com/SockDrawer/StoryBB into contextual-mode

Arantor committed:
Separate account and character groups, and force the main listing to point to the default profile.

Arantor committed:
Hide some of the character groups' permissions because they don't really apply to character groups.

Arantor committed:
Additional permissions in the character groups that should be hidden.

Arantor committed:
Don't apply the max-PM limit to character groups since it doesn't really make sense.

Arantor committed:
Make account vs character groups more prominent.

Arantor committed:
Don't let character groups inherit from account groups and vice versa.

Arantor committed:
Some code deduplication.

Arantor committed:
Enforce account/character inheritance on edit-group too.

Arantor committed:
Don't let character groups be converted into post count groups.

Arantor committed:
First stage of gluing it all together in the quick reply view.

Arantor committed:
Make sure we expose the correct ID to the quick reply even if there's only one choice.

Arantor committed:
Don't duplicate loading; the routine in Subs.php will make sure this is loaded if needed.

Arantor committed:
Rework implementation of IC/OOC to not have stupid edge cases.

Arantor committed:
No point checking this list twice.

Arantor committed:
Display some kind of hint if the person being posted as will require approval from quick reply.

Arantor committed:
Move the getting of the master list of possible characters into a generic function.

Arantor committed:
Typo fix

Arantor committed:
Don't let copying permissions from presets or new group creation give permissions to characters they can't otherwise have.

Arantor committed:
Merge branch 'master' of https://github.com/SockDrawer/StoryBB into contextual-mode

Arantor committed:
Apply filtering on whether user started the topic slightly differently because different calling points won't set the same global vars (yay)

Arantor committed:
Allow the character permission functions to filter on a specific character.

Arantor committed:
Fix an upstream bug where things were getting marked unapproved without any valid reason for it.

Arantor committed:
First pass of enforcing permissions for characters on posting.

Arantor committed:
Errors shouldn't be extra-escaped.

Arantor committed:
Hopefully this doesn't break anything.

Arantor committed:
Don't let guests have posting rights.

Arantor committed:
Make sure in the posting dropdown we also export the details of whether it needs approval to tell the user.

Arantor committed:
Handle editing. I think.

Arantor committed:
Don't show the topic creation buttons without proper permissions.

Arantor committed:
Merge branch 'master' into contextual-mode

Arantor committed:
Apply warning permissions removals during local cache for contextual mode.

(When we're satsfied this is correct, I'll add in a longer term cache mode for it as well that holds the profiles/boards/characters stuff but not the session/warning based checks to mitigate the additional performance hits.)

Fixes #465

Arantor committed:
Merge branch 'master' into contextual-mode

yamikuronue committed:
Merge branch 'master' into contextual-mode

yamikuronue commented on Sources/ManageMembergroups.php:
```diff
@@ -797,6 +825,31 @@ function EditMembergroup()
 	{
+		// Are we inheriting? Account groups can't inherit from character groups, and vice versa.
+		$current_group_is_character = StoryBB\Model\Group::is_character_group((int) $_REQUEST['group']);
+		if (isset($_POST['group_inherit']) && $_POST['group_inherit'] != -2)
```
negative two?

Arantor replied:
There's a bunch of ways that a group can be inherited or copied for permissions. Positive numbers indicate a group to inherit from, including -1 for guests. -2 relates to the specific type of inheritance here where we don't copy from the group but try to reuse it. It's weird and ugly and could do with a rethink sometime.

yamikuronue commented on Sources/Post.php:
```diff
@@ -559,25 +569,28 @@ function Post($post_errors = array())
 		$smcFunc['db_free_result']($request);
 
-		if ($row['id_member'] == $user_info['id'] && !allowedTo('modify_any'))
+		if ($row['id_member'] == $user_info['id'] && !charactersAllowedTo('modify_any'))
```
is `charactersAllowedTo` different than `areCharactersAllowedTo`?

Arantor replied:
Yes. allowedTo and charactersAllowedTo are soft checks, they return true/false. isAllowedTo and areCharactersAllowedTo are hard checks - failing them hits an error page directly (or login page for guests)

yamikuronue commented on Sources/Security.php:
```diff
@@ -1013,6 +1025,328 @@ function isAllowedTo($permission, $boards = null)
+ * @param string|array $permission A single permission to check or an array of permissions to check
+ * @param int|array $boards The ID of a board or an array of board IDs if we want to check board-level permissions
+ * @param int $character A specific character to filter by for easier perms checking.
+ * @return array Which character IDs are permitted to carry out the permission 
```
oh, disregard earlier comment. Blech on the similar naming but it's the existing pattern extended

Arantor replied:
Yup, it'll get renamed and revamped sometime.

Arantor committed:
Add a hopefully complete set of board access tests. (Needs a new Gherkin step because logging in has a rate limiter and we don't particularly want to bypass that.)

Arantor committed:
Merge branch 'master' into contextual-mode

Arantor committed:
Merge branch 'master' into contextual-mode

Arantor committed:
Fix #512

microtime() inexplicably returns '0.01234 1456789012' rather than a sane value. We only care about the fractional part here as it has more entropy.

Arantor committed:
Make group name a required field when playing with member groups. (Partial fix for #511)

Arantor committed:
Preserve state of character vs account if a group name isn't given. Fixes #511

Arantor committed:
Prevent a warning being shown for guests in immersive mode.

Arantor committed:
Merge branch 'master' of https://github.com/StoryBB/StoryBB into contextual-mode

Arantor committed:
Merge branch 'master' of https://github.com/StoryBB/StoryBB into contextual-mode

Arantor committed:
Merge branch 'master' of https://github.com/StoryBB/StoryBB into contextual-mode

Arantor committed:
Merge branch 'master' of https://github.com/StoryBB/StoryBB into contextual-mode

Arantor committed:
Merge branch 'master' of https://github.com/StoryBB/StoryBB into contextual-mode

# Conflicts:
#	Themes/default/languages/Admin.english.php
#	Themes/default/languages/Help.english.php

Arantor committed:
Merge branch 'master' of https://github.com/StoryBB/StoryBB into contextual-mode

Arantor committed:
Merge branch 'master' of https://github.com/StoryBB/StoryBB into contextual-mode

Arantor committed:
Merge branch 'master' of https://github.com/StoryBB/StoryBB into contextual-mode

# Conflicts:
#	other/behat/board_access.feature

Arantor committed:
Merge branch 'master' of https://github.com/StoryBB/StoryBB into contextual-mode

Arantor committed:
Merge branch 'master' of https://github.com/StoryBB/StoryBB into contextual-mode

# Conflicts:
#	Sources/Admin-Chars.php
#	Sources/Load.php
#	Themes/default/languages/Admin.english.php
#	Themes/default/languages/Help.english.php
